src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -583,6 +584,15 @@ where
                 }
                 (binding_ts.try_into().expect("must fit"), ingested_upper)
             };
+
+            // Changing `reclock_to_latest` can cause both the binding ts and the observed source
+            // upper to regress temporarily. We need to avoid producing a binding in that case.
+            if binding_ts <= prev_binding_ts
+                || PartialOrder::less_than(&cur_source_upper, &prev_source_upper)
```

Comment:
We only need the `prev_binding_ts` check here since the source frontier is fine to be regressed (the `mint` function will ignore it). This can only be violated by the `!reclock_to_latest` branch so I think the fix should move in there. Specifically instead of just taking the next available `binding_ts` I think the second if block needs to be this:

```rust
let mut new_probe = None;
while prev_probe >= new_probe {
    ticker.tick().await;
    // We only proceed if the source upper frontier is not the minimum frontier. This
    // makes it so the first binding corresponds to the snapshot of the source, and
    // because the first binding always maps to the minimum *target* frontier we
    // guarantee that the source will never appear empty.
    let upstream_frontier = ingested_upper
        .wait_for(|f| *f.frontier() != [FromTime::minimum()])
        .await
        .unwrap()
        .frontier()
        .to_owned();
    let now = (now_fn)();
    let mut probe_ts = now - now % timestamp_interval_ms;
    if (now % timestamp_interval_ms) != 0 {
        probe_ts += timestamp_interval_ms;
    }
    new_probe = Some(Probe {probe_ts, upstream_frontier});
}
new_probe.expect("always Some here");
```

Reply:
Ack, good to know!